src/Elmish.WPF/BindingData.fs

```diff
@@ -134,30 +135,114 @@ and internal AlterMsgStreamData<'model, 'msg, 'bindingModel, 'bindingMsg, 'dispa
 
 
-and internal BindingData<'model, 'msg> =
-  | BaseBindingData of BaseBindingData<'model, 'msg>
-  | CachingData of BindingData<'model, 'msg>
-  | ValidationData of ValidationData<'model, 'msg>
-  | LazyData of LazyData<'model, 'msg, obj, obj>
-  | AlterMsgStreamData of AlterMsgStreamData<'model, 'msg, obj, obj, obj>
+and internal BindingData<'model, 'msg, 't> =
+  | BaseBindingData of BaseBindingData<'model, 'msg, 't>
+  | CachingData of BindingData<'model, 'msg, 't>
+  | ValidationData of ValidationData<'model, 'msg, 't>
+  | LazyData of LazyData<'model, 'msg, obj, obj, 't>
+  | AlterMsgStreamData of AlterMsgStreamData<'model, 'msg, obj, obj, obj, 't>
 
 
 
 module internal BindingData =
 
+  module private MapT =
+
+    let baseCase (fOut: 't0 -> 't1) (fIn: 't1 -> 't0) =
+      function
+      | OneWayData d -> OneWayData {
+          Get = d.Get >> fOut
+        }
+      | OneWayToSourceData d -> OneWayToSourceData {
+          Set = fIn >> d.Set
+        }
+      | OneWaySeqData d -> OneWaySeqData {
+          Get = d.Get
+          CreateCollection = d.CreateCollection >> CollectionTarget.mapCollection fOut
+          GetId = d.GetId
+          ItemEquals = d.ItemEquals
+        }
+      | TwoWayData d -> TwoWayData {
+          Get = d.Get >> fOut
+          Set = fIn >> d.Set
+        }
+      | CmdData d -> CmdData {
+          Exec = d.Exec
+          CanExec = d.CanExec
+          AutoRequery = d.AutoRequery
+        }
+      | SubModelData d -> SubModelData {
+          GetModel = d.GetModel
+          CreateViewModel = d.CreateViewModel >> fOut
+          UpdateViewModel = (fun (vm,m) -> d.UpdateViewModel (fIn vm, m))
+          ToMsg = d.ToMsg
+        }
+      | SubModelWinData d -> SubModelWinData {
+          GetState = d.GetState
+          CreateViewModel = d.CreateViewModel >> fOut
+          UpdateViewModel = (fun (vm,m) -> d.UpdateViewModel (fIn vm, m))
+          ToMsg = d.ToMsg
+          GetWindow = d.GetWindow
+          IsModal = d.IsModal
+          OnCloseRequested = d.OnCloseRequested
+        }
+      | SubModelSeqUnkeyedData d -> SubModelSeqUnkeyedData {
+          GetModels = d.GetModels
+          CreateViewModel = d.CreateViewModel
+          CreateCollection = d.CreateCollection >> CollectionTarget.mapCollection fOut
+          UpdateViewModel = d.UpdateViewModel
+          ToMsg = d.ToMsg
+        }
+      | SubModelSeqKeyedData d -> SubModelSeqKeyedData {
+          GetSubModels = d.GetSubModels
+          CreateViewModel = d.CreateViewModel
+          CreateCollection = d.CreateCollection >> CollectionTarget.mapCollection fOut
+          UpdateViewModel = d.UpdateViewModel
+          ToMsg = d.ToMsg
+          VmToId = d.VmToId
+          BmToId = d.BmToId
+        }
+      | SubModelSelectedItemData d -> SubModelSelectedItemData {
+          Get = d.Get
+          Set = d.Set
+          SubModelSeqBindingName = d.SubModelSeqBindingName
+        }
```

Comment:
```suggestion
      | SubModelSelectedItemData d -> SubModelSelectedItemData d
```

Reply:
Same here, the only reason this works is because of a weird exception where the type of the specific item isn't actually held in the functions stored there. If that design ever changes, this will be more confusing to come across, not less.